wayland.c

```diff
@@ -96,6 +101,40 @@ static void destroy_output(struct mako_output *output) {
 	free(output);
 }
 
+static void touch_handle_motion(void *data, struct wl_touch *wl_touch,
+		uint32_t time, int32_t id,
+		wl_fixed_t surface_x, wl_fixed_t surface_y) {
+	struct mako_seat *seat = data;
+	seat->touch.x = wl_fixed_to_int(surface_x);
+	seat->touch.y = wl_fixed_to_int(surface_y);
+}
+
+static void touch_handle_down(void *data, struct wl_touch *wl_touch,
+		uint32_t serial, uint32_t time, struct wl_surface *sfc, int32_t id,
+		wl_fixed_t surface_x, wl_fixed_t surface_y) {
+	struct mako_seat *seat = data;
+	seat->touch.x = wl_fixed_to_int(surface_x);
+	seat->touch.y = wl_fixed_to_int(surface_y);
+}
+
+static void touch_handle_up(void *data, struct wl_touch *wl_touch,
+		uint32_t serial, uint32_t time, int32_t id) {
+	struct mako_seat *seat = data;
+	struct mako_state *state = seat->state;
+
+	struct mako_notification *notif;
+	wl_list_for_each(notif, &state->notifications, link) {
+		if (hotspot_at(&notif->hotspot, seat->touch.x, seat->touch.y)) {
+			// simulate a touch as a simple button press of type
+			// BTN_TOUCH
+			notification_handle_button(notif, BTN_TOUCH,
+				WL_POINTER_BUTTON_STATE_PRESSED);
+			break;
```

Comment:
i don't love this.

Reply:
Indeed. What about extracting most of `notification_handle_button`'s code into a new `notification_execute_binding(struct mako_notification *notif, enum mako_button_binding binding)` function?

Reply:
i think that works, updated! thanks.